odoo/tests/common.py

```diff
@@ -415,17 +415,20 @@ def with_user(self, login):
     @contextmanager
     def _assertRaises(self, exception, *, msg=None):
         """ Context manager that clears the environment upon failure. """
-        with super(BaseCase, self).assertRaises(exception, msg=msg) as cm:
+        with ExitStack() as init:
             if hasattr(self, 'env'):
-                with self.env.cr.savepoint():
-                    if issubclass(exception, AccessError):
-                        # The savepoint() above calls flush(), which leaves the
-                        # record cache with lots of data.  This can prevent
-                        # access errors to be detected. In order to avoid this
-                        # issue, we clear the cache before proceeding.
-                        self.env.cr.clear()
-                    yield cm
-            else:
+                init.enter_context(self.env.cr.savepoint())
+                if issubclass(exception, AccessError):
+                    # The savepoint() above calls flush(), which leaves the
+                    # record cache with lots of data.  This can prevent
+                    # access errors to be detected. In order to avoid this
+                    # issue, we clear the cache before proceeding.
+                    self.env.cr.clear()
+
+            with ExitStack() as inner:
+                cm = inner.enter_context(super().assertRaises(exception, msg=msg))
+                inner.push(init.pop_all())
```

Comment:
That's very clever, but I think it deserves a comment explaining that it makes the savepoint above exit before `assertRaises` does.